numpy/core/src/umath/simd.inc.src

```diff
@@ -3697,4 +3736,142 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
 
 #endif /* NPY_HAVE_SSE2_INTRINSICS */
 
+#ifdef NPY_HAVE_NEON
+/**begin repeat
+ *  #type = npy_float, npy_double#
+ *  #TYPE = FLOAT, DOUBLE#
+ *  #scalarf = npy_sqrtf, npy_sqrt#
+ *  #Nvtype128 = float32x4_t, float64x2_t#
+ *  #Nvtype64 = float32x2_t, float64x1_t#
+ *  #Nvpre = v, v#
+ *  #Nvsuf = f32, f64#
+ */
+
+
+/**begin repeat1
+* Arithmetic
+* # kind = add, subtract, multiply, divide#
+* # OP = +, -, *, /#
+* # VOP = add, sub, mul, div#
+*/
+
+static void
+neon_binary_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2, npy_intp n)
+{
+#ifdef  __aarch64__
+    const npy_intp vector_size_bytes = 16;
+    LOOP_BLOCK_ALIGN_VAR(op, @type@, vector_size_bytes)
+        op[i] = ip1[i] @OP@ ip2[i];
+    /* lots of specializations, to squeeze out max performance */
+    if (ip1 == ip2) {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype128@ a = @Nvpre@ld1q_@Nvsuf@(&ip1[i]);
+            @Nvtype128@ c = @Nvpre@@VOP@q_@Nvsuf@(a, a);
+            @Nvpre@st1q_@Nvsuf@(&op[i], c);
+        }
+    }
+    else {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype128@ a = @Nvpre@ld1q_@Nvsuf@(&ip1[i]);
+            @Nvtype128@ b = @Nvpre@ld1q_@Nvsuf@(&ip2[i]);
+            @Nvtype128@ c = @Nvpre@@VOP@q_@Nvsuf@(a, b);
+            @Nvpre@st1q_@Nvsuf@(&op[i], c);
+        }
+    }
+
+#elif defined __aarch32__
+    const npy_intp vector_size_bytes = 8;
+    LOOP_BLOCK_ALIGN_VAR(op, @type@, vector_size_bytes)
+        op[i] = ip1[i] @OP@ ip2[i];
+    /* lots of specializations, to squeeze out max performance */
+    if (ip1 == ip2) {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype64@ a = @Nvpre@ld1_@Nvsuf@(&ip1[i]);
+            @Nvtype64@ c = @Nvpre@@VOP@_@Nvsuf@(a, a);
+            @Nvpre@st1_@Nvsuf@(&op[i], c);
+        }
+    }
+    else {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype64@ a = @Nvpre@ld1_@Nvsuf@(&ip1[i]);
+            @Nvtype64@ b = @Nvpre@ld1_@Nvsuf@(&ip2[i]);
+            @Nvtype64@ c = @Nvpre@@VOP@_@Nvsuf@(a, b);
+            @Nvpre@st1_@Nvsuf@(&op[i], c);
+        }
+    }
+
+#endif
+    LOOP_BLOCKED_END {
+        op[i] = ip1[i] @OP@ ip2[i];
+    }
+}
+
+
+static void
+neon_binary_scalar1_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2, npy_intp n)
```

Comment:
scalar1 and scalar2 is too similar, can you extract a public function?

Reply:
used macros to combine both into one.